File: sdks/kotlin/src/main/kotlin/com/example_app/scanner/Webhook.kt
package com.example_app.scanner

import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Verifies an inbound scan webhook signature.
 *
 *   Header: X-Scanner-Signature: t=<unix>,v1=<hex>
 *           hex = hmac_sha256("<t>.<rawBody>", secret)
 *
 * Returns true on valid signature within `toleranceSeconds`. Throws
 * on missing / malformed headers so misconfiguration fails loudly.
 */
object ScannerWebhook {

    fun verifySignature(
        rawBody: String,
        header: String?,
        secret: String,
        toleranceSeconds: Long = 300,
    ): Boolean {
        requireNotNull(header) { "Missing X-Scanner-Signature header." }

        val parts = header.split(",").mapNotNull { kv ->
            kv.split("=", limit = 2).takeIf { it.size == 2 }
                ?.let { it[0].trim() to it[1].trim() }
        }.toMap()

        val t = parts["t"]?.toLongOrNull()
            ?: throw IllegalArgumentException("Malformed signature header (no t).")
        val v1 = parts["v1"]
            ?: throw IllegalArgumentException("Malformed signature header (no v1).")

        val now = System.currentTimeMillis() / 1000
        if (kotlin.math.abs(now - t) > toleranceSeconds) return false

        val mac = Mac.getInstance("HmacSHA256").apply {
            init(SecretKeySpec(secret.toByteArray(Charsets.UTF_8), "HmacSHA256"))
        }
        val expected = mac.doFinal("$t.$rawBody".toByteArray(Charsets.UTF_8))
            .joinToString("") { "%02x".format(it) }

        return constantTimeEquals(expected, v1)
    }

    private fun constantTimeEquals(a: String, b: String): Boolean {
        if (a.length != b.length) return false
        var diff = 0
        for (i in a.indices) {
            diff = diff or (a[i].code xor b[i].code)
        }
        return diff == 0
    }
}

File: sdks/swift/Sources/ExampleAppScannerSDK/Webhook.swift
import Foundation
import CryptoKit

/// Verifies an inbound scan webhook signature.
///
/// Header shape from the platform:
///   `X-Scanner-Signature: t=<unix>,v1=<hex>` where
///   `hex = hmac_sha256("<t>.<rawBody>", secret)`.
///
/// Returns true when the signature is valid and within `tolerance`
/// of the current time. Throws on missing / malformed headers so
/// misconfigured backends fail loudly.
public enum ScannerWebhook {

    public static func verifySignature(
        rawBody: String,
        header: String?,
        secret: String,
        tolerance: TimeInterval = 300
    ) throws -> Bool {
        guard let header else { throw NSError(domain: "ExampleAppScannerSDK", code: 1, userInfo: [NSLocalizedDescriptionKey: "Missing X-Scanner-Signature"]) }

        var parts: [String: String] = [:]
        for kv in header.split(separator: ",") {
            let halves = kv.split(separator: "=", maxSplits: 1).map(String.init)
            guard halves.count == 2 else { continue }
            parts[halves[0].trimmingCharacters(in: .whitespaces)] = halves[1].trimmingCharacters(in: .whitespaces)
        }
        guard let t = parts["t"], let v1 = parts["v1"], let ts = TimeInterval(t) else {
            throw NSError(domain: "ExampleAppScannerSDK", code: 2, userInfo: [NSLocalizedDescriptionKey: "Malformed signature header"])
        }
        if abs(Date().timeIntervalSince1970 - ts) > tolerance {
            return false
        }

        let payload = "\(Int(ts)).\(rawBody)"
        let key = SymmetricKey(data: Data(secret.utf8))
        let mac = HMAC<SHA256>.authenticationCode(for: Data(payload.utf8), using: key)
        let hex = mac.map { String(format: "%02x", $0) }.joined()
        return constantTimeEquals(hex, v1)
    }

    private static func constantTimeEquals(_ a: String, _ b: String) -> Bool {
        guard a.count == b.count else { return false }
        var diff: UInt8 = 0
        for (x, y) in zip(a.utf8, b.utf8) { diff |= x ^ y }
        return diff == 0
    }
}
